### DISCUSSION TITLE
CI (GitHub Actions)

### DISCUSSION THREAD

**Discussion by esensar** [FossifyOrg/General-Discussion#83] [Ideas]
Hello :wave: 

I have been an active (paid) contributor to SimpleMobileTools for a couple of months last year and I believe that not having any kind of automated checks (and no automatic formatting) caused unnecessary friction. PRs were often sitting for quite some time, only for the sole maintainer to notice some minor formatting issue or something like that and then waiting for more time for the PR creator to fix it.

I also believe it would help new contributors be more confident that their solution is not breaking some of the rules of the project. I would gladly take this up and set it up for all of the applications if it sounds like a good idea.

**Comment by naveensingh** [comment]
@esensar Hey, that sounds like a good idea. PRs are welcome :)

**Reply by esensar** [reply]
I have created a sample PR: https://github.com/FossifyOrg/Launcher/pull/12
I guess we can use it to define some common rules and figure out how to handle transition to that workflow (since we will start with a bunch of issues, since we haven't used any of these tools). I have also included `test` phase in the workflow, even though no tests are written for any of the apps. It will always pass right now. I can remove it if we definitely don't want any tests, but I think covering some of the core parts could be beneficial.

**Reply by esensar** [reply]
We should also probably decide if we want to immediately fix issues reported by these tools right now, or just add current issues to baseline and ensure that no more issues are created on future PRs.

**Reply by inson1** [reply]
If we choose first option it would take longer to get it working, right? Because for sure there lots of issues. So I'm for creating issue for cleaning it up (in general issues with nice table of the repostories apps) and just start using it.

**Reply by naveensingh** [reply]
@esensar 

> just add current issues to baseline and ensure that no more issues are created on future PRs.

yes, please let's add current issues to the baseline.

>  I think covering some of the core parts could be beneficial.

I agree. We'll add tests in the future as right now the focus is to just keep the project alive by squashing bugs and implementing important features.

**Reply by naveensingh** [reply]
We could also automatically build an APK when a PR is created so that people can easily test it out.

**Reply by esensar** [reply]
That is a good idea. Should we make it on every PR or make it manually triggerable or something like that? I am thinking something like a label to enable that on a PR.
I am using that first PR to iron out such things, so that I could just copy it to other repositories when it's done.

**Reply by esensar** [reply]
Actually, we could also use [reusable workflows](https://docs.github.com/en/actions/using-workflows/reusing-workflows) (we could maybe put them in Commons). I will finalize this PR first and then later on we could go for reusable ones.

**Reply by naveensingh** [reply]
> Should we make it on every PR or make it manually triggerable or something like that? I am thinking something like a label to enable that on a PR.

We need to disable this at least for Weblate PRs as they are like a recurring theme and don't need to be tested. If you want to enable it only for a specific label, let me know and I'll add another label like "testers needed" or something like that to all repos. Doesn't seem like we need to think about [usage limits](https://docs.github.com/en/actions/learn-github-actions/usage-limits-billing-and-administration#usage-limits) yet. 

Regarding your other comment about the APK being hidden, I agree. We can maybe use [this action](https://github.com/marketplace/actions/comment-pull-request) to comment a link to the APK. I'll do it later but If you want to do it, feel free to do it in the existing Launcher PR. 

Once we are done sorting this out, we can go for reusable workflows in Commons.

**Reply by naveensingh** [reply]
The [.github repo](https://github.com/FossifyOrg/.github) is another (better?) place for reusable workflows but I'm not sure if GitHub has any restrictions around it as it's a special repository.

**Reply by esensar** [reply]
Yeah, that works too. I think it is even recommended (and can also host custom actions, not just workflows).

**Reply by esensar** [reply]
I have updated the workflow to make build only when 'testers needed' label is on the PR (it is also triggered if label has been added later on).

Could you add that label to that PR when you get some time, so we can just check if the workflow works as expected? After that I think we are done for that PR.

**Reply by naveensingh** [reply]
Added the label but the comment failed: https://github.com/FossifyOrg/Launcher/actions/runs/7671051341/job/20908595552?pr=12

**Reply by naveensingh** [reply]
Granting the write permission should solve it as per https://github.com/marketplace/actions/comment-pull-request#permissions

**Reply by esensar** [reply]
Looks like write permission can't be granted for forked repositories: https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#workflows-in-forked-repositories

I guess we will have to manually remind people or find some other way to lead people to the artifact

**Reply by naveensingh** [reply]
Actually, I'm detaching the forks, it might take a day or two as the GitHub staff will do it manually.

**Reply by esensar** [reply]
I think this refers to the fact that my repository is forked from Fossify and running workflows on that PR would be unsafe for write permission. I think even after detaching, it will still be an issue, since most of the PRs will come from forks and not directly from the FossifyOrg organization.

**Reply by naveensingh** [reply]
I guess I'll merge it now then. I'll investigate this later.

**Reply by esensar** [reply]
I have now added reusable workflows: https://github.com/FossifyOrg/.github/pull/7

There are also templates there, which will make it easier to use these workflows.

**Reply by esensar** [reply]
By the way, I found an examle of commenting workflow. To get around the permission issue, it is done as a separate workflow, triggered by the completion of PR workflow: https://github.com/bevyengine/bevy/blob/main/.github%2Fworkflows%2Fci-comment-failures.yml

We could maybe consider something similar.

**Reply by naveensingh** [reply]
That solution looks promising. I tried using `pull_request_target` instead of `pull_request` in the [image-minimizer here](https://github.com/FossifyOrg/Messages/blob/master/.github/workflows/image-minimizer.yml) but it doesn't seem to work: https://github.com/FossifyOrg/Messages/actions/runs/7677669075/job/20930629270?pr=53

From [this page](https://github.com/marketplace/actions/comment-pull-request#permissions):
>Note that, if the PR comes from a fork, it will have only read permission despite the permissions given in the action for the pull_request event. In this case, you may use the pull_request_target event. With this event, permissions can be given without issue (the difference is that it will execute the action from the target branch and not from the origin PR).

But with these, we need to make sure to never check out untrusted code: https://nathandavison.com/blog/github-actions-and-the-threat-of-malicious-pull-requests

**Reply by esensar** [reply]
Yeah, we would have to move away from gradle wrapper in that case, since it is a local script that could be rewritten and do whatever on that PR, while having full access. We would have to use that checkout to get code from the PRs branch, which would leave us vulnerable. I think that workflow complete hook is safer.

**Comment by naveensingh** [comment]
Thanks a lot for the help, closing this now.
